cupy/core/core.pyx

```diff
@@ -1990,6 +1996,323 @@ cpdef ndarray dot(ndarray a, ndarray b, ndarray out=None):
     return tensordot_core(a, b, out, n, m, k, ret_shape)
 
 
+cpdef ndarray _get_all_addresses(size_t start_adr,
+                                 vector.vector[size_t]& shape,
+                                 vector.vector[size_t]& strides):
+    idx = numpy.array([start_adr])
+    for sh_, st_ in zip(shape, strides):
+        idx = (idx[:, None] + (numpy.arange(sh_) * st_)[None, :]).ravel()
+    idx = idx.astype(numpy.uintp)
+
+    ret = ndarray((idx.size,), dtype=numpy.uintp)
+    ret.set(idx)
+    return ret
+
+
+cdef ndarray _mat_ptrs(ndarray a):
+    """Creates an array of pointers to matrices
+    Args:
+        a: A batch of matrices on GPU.
+           shape: () -> one ptr
+           shape: (A) -> one ptr to mat o size (A)
+           shape: (A, B) -> one ptr to mat o size (A, B)
+           shape: (A, B, C) -> A ptrs to mat o size (B, C)
+           shape: (A_1, ..., A_N, B, C) -> A_1*...*A_N ptrs to mat of
+                  size (B, C)
+    Returns:
+        GPU array of pointers to matrices.
+    """
+    cdef Py_ssize_t stride, ptr, pointer, i
+    cdef ndarray ret
+    if a.ndim <= 2:
+        ret = ndarray((1,), dtype=numpy.uintp)
+        ret.fill(a.data.ptr)
+        return ret
+    else:
+        return _get_all_addresses(a.data.ptr, a.shape[:-2], a.strides[:-2])
+
+
+cpdef ndarray matmul(ndarray a, ndarray b):
+    """
```

Comment:
Please write one line document for function.
For example `Matrix product of two arrays`.